### PR TITLE
adds support for pulling social data (twitter + bitcointalk + reddit )

### PR DESCRIPTION
**Pull request by owocki**
# wat

* twitter
* bitcointalk
* reddit

Searches the above social networks every couple minutes for new tweets about currencies listed in `settings.SOCIAL_NETWORK_SENTIMENT_CONFIG`.  Saves the post body, a rudimentary sentiment analysis, and a handful of other metadata to a new model called `SocialNetworkMention`.

twitter requires the following additions to `local_settings.py`:

```
# only required for pull_twitter.py
# get this info @ https://apps.twitter.com/
TWITTER_CONSUMER_KEY = ''
TWITTER_CONSUMER_SECRET = ''
TWITTER_ACCESS_TOKEN_KEY = ''
TWITTER_ACCESS_TOKEN_SECRET = ''
```

you may also want to override this in local_settings.py:
```
SOCIAL_NETWORK_SENTIMENT_CONFIG = {
    'twitter': ['ETH', 'BTC', 'FCT', 'BTS', 'XMR', 'MAID', 'DASH'],
    'bitcointalk': ['ETH', 'BTC', 'FCT', 'BTS', 'XMR', 'MAID', 'DASH'],
    'reddit': {
        'ethereum': ['ETH'],
        'ethtrader': ['ETH'],
        'dogecoin': ['DOGE'],
        'bitcoin': ['BTC'],
        'bitcoinmarkets': ['BTC'],
    }
}
```

Example of admin having data:

# Img

### CODE REVIEW

File: history/models.py [hunk 1]
```diff
@@ -563,3 +563,12 @@ def predict(self, sample):
         recommend = self.recommend_trade(nn_price, last_sample)
         projected_change_pct = (nn_price - last_sample) / last_sample
         return recommend, nn_price, last_sample, projected_change_pct
+
+
+class SocialNetworkMention(AbstractedTesterClass):
+    network_name = models.CharField(max_length=30, db_index=True)
+    network_username = models.CharField(max_length=100)
+    network_id = models.BigIntegerField(default=0)
+    network_created_on = models.DateTimeField(default=get_time)
```
Comment (rubik) [inline history/models.py]: This should be `auto_now_add=True` instead of default. `get_time` is not UTC and I left it there only because it's used in an old migration.

**Comment by rubik** [comment]
Wow, great work!

**Comment by rubik** [comment]
Do you think it makes sense to add basic sentiment analysis in the model? I don't know, something like

    import textblob
    textblob.TextBlob(tweet.text).sentiment.polarity

Adding it right into the DB would make it easier to graph/inspect later.

**Comment by owocki** [comment]
reddit support added in 42ae0cdcad35a14a61ed41293f325590b59613a0

File: history/management/commands/pull_bitcointalk.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            children = {child.tag: child.text.encode('utf-8') for child in item}
+            post_link = children['link']
+            response = requests.get(post_link)
+            response_body = response.text.encode('utf-8')
+            parsed_html = BeautifulSoup(response_body)
+            post_body = parsed_html.find('div', attrs={'class': 'post'}).text
+            message_id = get_message_id(children['guid'])
+            tmp_date = children['pubDate'].split(' ')  # Sat, 16 Apr 2016 18:23:51 GMT
+            tmp_time = tmp_date[4].split(':')  # 18:23:51
+            network_created_on = datetime.datetime(int(tmp_date[3]), int(monthToNum(tmp_date[2])), int(tmp_date[1]), int(tmp_time[0]), int(tmp_time[1]), int(tmp_time[2]))
```
Comment (igorpejic) [inline history/management/commands/pull_bitcointalk.py]:
```python
network_created_on = datetime.datetime.strptime(children['pubDate'], "%a, %d %b %Y %X %Z")
```

**Comment by owocki** [comment]
> Do you think it makes sense to add basic sentiment analysis in the model? I don't know, something like


@rubik awesome idea!  21386d3792f16bb459989b0950e78d2336b8aed8

**Comment by owocki** [comment]
any other suggestions for social networks to analyze?  reddit/bitcointalk/twitter feel like they'd be enough to give initial signal.

shameless asks: 
1. it might be worth having a shared server that collects this information and nightly dumps.  @Snipa22 any interest in hooking this up? 
1. who wants to help me feed this into the NNs and figure out how to get alpha ? :)

File: history/management/commands/pull_bitcointalk.py (hunk 2 above)
Comment (owocki) [inline history/management/commands/pull_bitcointalk.py]: doh, i knew there was an easier way to do this.  didnt see the `%a` option.

**Comment by rubik** [comment]
> 1. who wants to help me feed this into the NNs and figure out how to get alpha ? :)

I'm interested, but I have close to zero experience with NNs (just a bit of theory). I've only worked with sklearn classifiers. What would this entail? If I am not mistaken alpha is the parameter that controls gradient descent.

**Comment by owocki** [comment]
> If I am not mistaken alpha is the parameter that controls gradient descent.

I meant alpha, as defined in finance: "The excess returns of a fund relative to the return of a benchmark index is the fund's alpha."  http://www.investopedia.com/terms/a/alpha.asp

I wrote up some documentation on how this repo is used and how to trade @ https://github.com/owocki/pytrader/pull/40

File: history/management/commands/pull_bitcointalk.py (hunk 2 above)
Comment (owocki) [inline history/management/commands/pull_bitcointalk.py]: e4fd13d6f4ad0102ba6cc83e86b6e86c933bd031

File: history/models.py (hunk 1 above)
Comment (owocki) [inline history/models.py]: 6eb1b033fbdcabde8d0af3f9b00988304af57d52

**Comment by owocki** [comment]
@rubik just sent you a few books on ML via slack. dont feel obligated, but if you're interested....

**Comment by rubik** [comment]
> I meant alpha, as defined in finance: "The excess returns of a fund relative to the return of a benchmark index is the fund's alpha." http://www.investopedia.com/terms/a/alpha.asp

Oh, now it's clearer. And thanks for the books, definitely interested!

File: history/management/commands/pull_reddit.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            ]
+            for submissions in submission_set:
+                for x in submissions:
+                    network_created_on = datetime.datetime.fromtimestamp(x.created_utc)
+                    if SocialNetworkMention.objects.filter(network_name='reddit', network_id=x.id).count() == 0:
+                        for currency_symbol in currencies:
+                            snm = SocialNetworkMention.objects.create(
+                                network_name='reddit',
+                                network_id=x.id,
+                                network_created_on=network_created_on,
```
Comment (rubik) [inline history/management/commands/pull_reddit.py]: Unfortunately this does not work. When `auto_now_add=True`, Django will override whichever value you pass. Not very intuitive, I know. We have three options:

* temporarily disable `auto_now_add`, saving, and enable it again. This can be done with a custom context manager ([example](http://stackoverflow.com/a/35943149/448496));
* changing the value after saving and saving again -- probably the worst choice;
* replacing `auto_now/auto_now_add` with a default like `timezone.now` and overloading the `save` method.

I think that if a custom datetime is needed only for this case the first approach could be viable. Otherwise, the best one is the third. If we choose the third one I can add this change to my PR, it's not much of a big deal.

File: history/management/commands/pull_bitcointalk.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+from django.core.management.base import BaseCommand
+from django.conf import settings
+from history.models import SocialNetworkMention
+
+import xml.etree.ElementTree as ET
+from BeautifulSoup import BeautifulSoup
+
+
+def monthToNum(date):
```
Comment (igorpejic) [inline history/management/commands/pull_bitcointalk.py]: Is this function used anywhere?

File: history/management/commands/pull_reddit.py (hunk 3 above)
Comment (owocki) [inline history/management/commands/pull_reddit.py]: `network_created_on != created_on` -- the former is the date I receive from the other network, the latter is the creation date of this object in the DB.

File: history/management/commands/pull_bitcointalk.py (hunk 4 above)
Comment (owocki) [inline history/management/commands/pull_bitcointalk.py]: not anymore cae6fea05bf8dc5b4faf8190f0d733a8bef96470

File: history/management/commands/pull_reddit.py (hunk 3 above)
Comment (owocki) [inline history/management/commands/pull_reddit.py]: doh, i now see i added the `auto_now_add` to the wrong field.  updating..

Comment (rubik) [inline history/management/commands/pull_reddit.py]: Yeah, I didn't see that the timestamp is always available. Without `auto_now_add` is much simpler.

**Comment by jeff-hykin** [comment]
Hey Owocki!

Been awhile, but I’m finally getting a little time to work on this. Amazing job pulling the semantics, that was exactly what I was hoping to do. I still haven’t had time to even get code running on the raspberry pi, but I still intent to keep working on the goals I setup in issue #20 . 

My suggestion for grabbing more info would be to add journal articles and media. For example scrape Twitter for URLs that link to crypto coin keywords, the more tweets that include link to the same URL, it is probably more likely the URL's info will affect the actual market price. 

There are more things I’d like to address, but I’ve got to go for now.

**Comment by jeff-hykin** [comment]
Check this out [www.hedgechatter.com](url) They’re basically doing/have done what this issue is trying to accomplish 

Sadly I’ve had little free time due to work. I still haven’t even run a single line of code on my Ras-pi. The good news is I’ve been learning Python at work that should help the implementation of this project.
